telegram/src/main/kotlin/com/github/kotlintelegrambot/dispatcher/handlers/InputChainHandler.kt

```diff
@@ -0,0 +1,65 @@
+package com.github.kotlintelegrambot.dispatcher.handlers
+
+import com.github.kotlintelegrambot.Bot
+import com.github.kotlintelegrambot.entities.Message
+import com.github.kotlintelegrambot.entities.Update
+import com.github.kotlintelegrambot.extensions.filters.Filter
+
+data class InputChainHandlerEnvironment<T>(
+    val inputChainIdGen: (previousInputChainId: T?, Update) -> T?,
+    val inputChainId: T,
+    val bot: Bot,
+    val update: Update,
+    val message: Message,
+)
+
+class InputChainHandler<T>(
+    private val filter: Filter,
+    private val initialHandler: Handler,
+    private val inputChainIdGen: (previousInputChainId: T?, Update) -> T?,
+    private val handleInputChain: HandleInputChain<T>,
+) : Handler {
+    private val chainMap = mutableMapOf<Long, T?>()
+
+    override fun checkUpdate(update: Update): Boolean =
+        initialHandler.checkUpdate(update) || update.check()
+
+    override suspend fun handleUpdate(bot: Bot, update: Update) {
+        update.message?.let { message ->
+            message.from?.id?.let { userId ->
+                chainMap.remove(userId)?.let { inputChainId ->
+                    handleInputChain(
+                        InputChainHandlerEnvironment(
+                            inputChainIdGen,
+                            inputChainId,
+                            bot,
+                            update,
+                            message,
+                        ),
+                    )?.let { chainMap[userId] = it }
+                    Unit
+                }
+            }
+        } ?: initialHandler.handleUpdate(bot, update).also {
+            update.takeIf { it.consumed }
```

Comment:
Why only chaining if the update is consumed? I feel that having to mark the first update as consumed for the chain to work is going to make the API hard to use.

Reply:
I was thinking about it...
I assumed whenever "a question" is processed, they (App devs) should consume it to avoid wrong behavior in other commands that match the "initialHandler" checks (And maybe override the ID stored on this chain), but maybe is a good idea removing this check and rely on the "good implementation" of or users